clang/test/SemaSYCL/sycl-external-attr.cpp

```diff
@@ -0,0 +1,70 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsycl-is-device -std=c++20 -fsyntax-only -verify -DCPP20 %s
+// Semantic tests for sycl_external attribute
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+static void func1() {}
+
+namespace {
+  [[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func2() {}
+
+  struct UnnX {};
+}
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
+// The first declaration of a SYCL external function is required to have this attribute.
+int foo(); // expected-note {{previous declaration is here}}
+
+[[clang::sycl_external]] int foo(); // expected-error {{'clang::sycl_external' attribute does not appear on the first declaration}}
```

Comment:
Avoid blank lines between declarations that are part of the same test so that it is easy to see that they go together. Use line markers and references in expected diagnostics to avoid line length issues.
```suggestion
// The first declaration of a SYCL external function is required to have this attribute.
// expected-error@+3 {{'clang::sycl_external' attribute does not appear on the first declaration}}
// expected-note@#func8-decl {{previous declaration is here}}
int func8(); // #func8-decl
[[clang::sycl_external]] int func8();
```